### ISSUE TITLE
[css-display] math/inline-math

### ISSUE THREAD

**Issue by bkardell** [w3c/csswg-drafts#5385]
Mathematical rendering, like other text renderings can happen inline with prose or in the block direction and different elements employ different strategies for layout.  According to common norms, these are treated a little differently.  [MathML Core proposes the introduction of two display types](https://mathml-refresh.github.io/mathml-core/#new-display-math-value):
  
`display: math` and `display: inline-math`

These control box generation and layout according to their tag name.  It shoud be possible for authors to experiment/extend and provide custom layouts for when authors would like to override.  The figure below illustrates differences in standard treatments of mathematical rendering depending on whether it is inline with text, or has a complete space in the block direction, as illustrated in the figure below. 

![illustration of an equation rendered inline with prose and again as a block](https://bkardell.com/media/mathml-inline-and-block.png)

These can be overridden and extended with custom layouts.  `display: contents` computes to ``none`.

**Comment by SebastianZ** [comment]
I'd say, the `math` keyword should be a [`<display-inside>`](https://drafts.csswg.org/css-display-3/#inner-model) value, so instead of `inline-math` it would be `inline math`.

Sebastian

**Comment by Loirooriol** [comment]
Not completely opposed to `inline-math`, but note we have the precedent of dropping `inline-list-item` in favor of the space-separated `inline list-item`.

**Comment by fred-wang** [comment]
The resolution says 'math' outside of MathML context computes to 'flow' but it seems that would mean outer display is block:

> If a <display-inside> value is specified but <display-outside> is omitted, the element’s outer display type defaults to block—except for ruby, which defaults to inline.

which is a bit different to what we discussed. I think we meant "math" and "inline-math" computes to "display flow" or "inline flow" respectively.

**Comment by fred-wang** [comment]
Note sure that's an argument for introducing the legacy display inline-math but according to MDN, Firefox is the only browser supporting multiple display value: https://developer.mozilla.org/en-US/docs/Web/CSS/display-inside#Browser_compatibility

`<div style="display: inline flow"></div>` has computed style "inline" in Firefox but "block" in Chrome.

**Comment by astearns** [comment]
My interpretation was that `display-inside:math` is what computes to `flow` outside of a MathML context. So `display: inline math` would compute to `display: inline flow`. But @fantasai was translating what I actually said into what made sense to her as she minuted - is this what you meant me to mean, @fantasai?

If that's correct, then I think it follows that `display:inline-math` computes to `display:inline flow` in that context.

**Comment by fred-wang** [comment]
Another note we discussed with @emilio : do we want to add a special case so that the style of a computed display value `inline math` (or `inline math`) serializes to `inline-math` rather than `inline math`?

**Comment by MatsPalmgren** [comment]
I don't think we should add `inline-math`. There's no reason to have that if we add `inline math`.

Given that `<math>` by default behaves as an inline in an inline context, it seems more logical that `display:math` should be the inline variant and `display:block math` should be the block variant, similar to how `ruby/block ruby` is designed.  Modeling it on an existing similar feature is likely easier to implement and spec,  and easier to learn for authors.

**Comment by Loirooriol** [comment]
@MatsPalmgren I think that's a very good point which I don't see discussed in the minutes. Agenda+ to discuss treating `display: math` as `display: inline math` for consistency with ruby.

**Comment by fred-wang** [comment]
Hi, another case that was not very explicit is the case of pseudo elements:

      ::before { content: 'math'; display: math; }
      ::after { content: 'math'; display: block math; }

since they are not MathML elements, I understand they also compute to inline/block flow.

**Comment by faceless2** [comment]
That seems a shame - if the engine is able to handle pseudo-elements within the Math context, they should be usable, for (eg) equation numbering. Could we redefine non-math elements as "non-pseudo elements in a namespace other than math"?

**Comment by fred-wang** [comment]
I don't know if there is a right solution, but from an implementation point of view, since these math values are going to allow new combinations and potentially lead to security bugs, I'd prefer to minimize the change if there is not a strong use case.

Note that equation numbering or any other text-only pseudo element can be done with display: block (or other non-math displays), the MathML layout should still be able to handle these non-math displays anyway.

Also, if we want to allow math display on pseudo elements, probably MathML Core will have to say pseudo-elements are laid out as an mrow and the spec updated at each place an "mrow-like" behavior is assumed ; adding special cases for pseudo-elements, which means more spec/test/implementation work. I'm personally not very willing to do this extra work for level 1.

**Comment by MatsPalmgren** [comment]
> but from an implementation point of view, since these math values are going to allow new combinations and potentially lead to security bugs

I think it's the opposite actually - making some values need special handling is usually what complicates implementations.  It's also harder to spec, and for authors to learn, a bunch of special rules for a _subset_ of a property's values rather than having the property  behave in a uniform way for all its values.

These new math `display` values should apply to all (pseudo-)elements exactly the same as other `display` values.  So, `<math>`  is from a box construction point of view exactly the same as `<div style="display:math">`, and conversely, `<math style="display:block">` is exactly the same as a plain `<div>`.  The UA sheet should have a `math { display: math }` rule that the author can override as they see fit (same as we do for `<ruby>`).

**Comment by fred-wang** [comment]
@MatsPalmgren `<math style="display:block">` is exactly the same as `<div>` but `<div style="display:math">` is *not* the same as `<math>` since we agreed that non-MathML elements would compute to flow (this is easy to do in the style adjuster). So you might not like it, but with the current spec and browser MathML implementations, having this exception for pseudo-elements too does not add complexity. On the other hand, allowing pseudo-element mrow definitely adds a new thing to handle in the implementations and for the spec / tests (perhaps not difficult but again I'm still waiting for a strong justification here).

BTW I disagree with the justification in https://github.com/w3c/csswg-drafts/issues/5385#issuecomment-692147232 ; for Chromium's implementation trying to align on ruby's model does not help at all for the MathML implementation, we can either choose to interpret `display: math` as block (like grid, flex...) or as inline (like ruby), that's a choice in the CSS parser that does not affect the MathML layout code. I think it's ok to choose inline since MathML3 says `<math>` is inline while block need a special `display="block"` attribute.

In general, I don't think trying to absolutely following `<ruby>` is correct here. MathML layout is much more complex and there are too many tag-based rules inherited from MathML3 ; we are not proposing to introduce a new display for each MathML tag or even just for each "kind of layout" (which could be controversial anyway since things like `display: fraction` would imply mathematical semantics instead of being just pure style).

**Comment by fantasai** [comment]
@fred-wang I agree with the logic in @MatsPalmgren's [comment](https://github.com/w3c/csswg-drafts/issues/5385#issuecomment-697697234), and it doesn't seem unreasonable to just say that "unless otherwise specified, a `math` display type makes any element behave like an mrow".  That's what happens to unknown elements in the MathML namespace already, right? So we would just extend the logic to all elements that don't have special math behaviors.  Would there be a particular problem with doing that?

**Comment by fred-wang** [comment]
@fantasai yes we can do that, which means we have to extend the MathML Core spec to treat all these new mrow-like like an mrow so this is indeed adding more work in the implementation/tests/spec (compared to computation to flow) ; not sure how difficult it will be or whether it will have unexpected security consequences, but for me it's safest to keep things simple for a first implementation.

**Comment by fantasai** [comment]
@fred-wang It's hardly anything in terms of spec work, not more complicated than saying it computes to `flow`. And in terms of tests, we need a test for what happens with these display values on non-MathML elements either way -- proving that it renders as `flow` rather than `mrow` is roughly equivalent to proving that it renders as `mrow` rather than `flow`. As for implementations, in Gecko at least I believe it's actually easier to map a `display` value to a box type unconditionally than to make it compute to a different value depending on its namespace/ancestry and I'd be surprised if the same isn't true for WebKit and Blink. The `<math>` element itself is following the same codepath we're proposing for `<div>` so by making them the same, we are not adding any new code path, only removing the exception for elements outside of MathML context.

**Comment by MatsPalmgren** [comment]
I wrote a [patch](https://bugzilla.mozilla.org/show_bug.cgi?id=1664261) that implement this in Gecko and it was pretty straight-forward. Gecko already supports multi-keyword `display` and `block ruby` of course, so that made it quite trivial to add similar code to handle `math/block math`. This demonstrates that @fred-wang's claims that it would be hard to implement or add complexity or that it would cause (unspecified) security issues are false.

**Comment by MatsPalmgren** [comment]
@fantasai wrote:
> And in terms of tests, we need a test for what happens with these display values on non-MathML elements either way

Indeed, and we should probably also move any spec text from the MathML spec to the CSS Display spec about any box tree "fixups" that happens when non-MathML elements are injected into this context.

For the record, here's what Gecko currently does:
https://searchfox.org/mozilla-central/rev/89d33e1c3b0a57a9377b4815c2f4b58d933b7c32/layout/base/nsCSSFrameConstructor.cpp#3764-3783
https://searchfox.org/mozilla-central/rev/89d33e1c3b0a57a9377b4815c2f4b58d933b7c32/layout/base/nsCSSFrameConstructor.cpp#4666-4702
https://searchfox.org/mozilla-central/rev/89d33e1c3b0a57a9377b4815c2f4b58d933b7c32/layout/mathml/mathml.css#213-217

Translated into English: each consecutive sequence of inlines or text are wrapped in an anonymous `-moz-mathml-anonymous-block` pseudo box, which is essentially a plain inline-block.  So for example:
```html
<math><ms>X</ms></math>
<script>
let math = document.querySelector("math");
let span = document.createElement("span");
span.append(document.createTextNode("b"));
math.prepend(span);
math.prepend(document.createTextNode("a"));
math.append(document.createTextNode("c"));
</script>
```

will create a box tree that looks something like this:
```
Inline(math) <
  ::-moz-mathml-anonymous-block(math) <
    Text("a")
    Inline(span) <
      Text("b")
    >
  >
  Box(ms)<... details intentionally left out ...>
  ::-moz-mathml-anonymous-block(math) <
    Text("c")
  >
>
```

which renders as: `ab"X"c`

I think details like this needs to be specced somewhere.

**Comment by MatsPalmgren** [comment]
BTW, the above fixups seems a bit odd to me.  I wonder if it would be better to use [inlinification](https://drafts.csswg.org/css-display/#inlinify) as we do for [`display:ruby`](https://drafts.csswg.org/css-ruby-1/#box-fixup) to consolidate on one set of rules for consistency?

**Comment by MatsPalmgren** [comment]
@fred-wang wrote:
> MathML layout is much more complex and there are too many tag-based rules inherited from MathML3 ; we are not proposing to introduce a new display for each MathML tag

There are only a dozen or so MathML-specific box types in Gecko, so I don't think that's a problem. CSS `display` can easily handle a dozen more `math-*` keywords if we need them.  Some elements like `<mtable>, <mtr>, <mtd>` can use already existing table values.

I discussed this with @emilio and we don't like that `display: math` should magically compute to different values based on the tag name or that it should magically create different box types for `math` based on the tag name. If we want to express MathML box construction in CSS then it should follow normal CSS rules and **not** have special MathML hacks for how property values are computed and applied.  Hacks like that only make CSS harder to implement, spec and teach to authors.

Fwiw, here's the MathML-specific box types that Gecko implements (besides `math/block math`): `token, mrow, mo, mfrac, mmultiscripts, munderover, mpadded, mspace, msqrt, mroot, maction, menclose, semantics`. (I don't know how well these match MathML Core. Maybe some could be merged if they are reasonably close in their layout?)

Anyway, we would prefer to keep `display` values having consistent behavior, i.e. that the value alone determines the box type to create, so we propose to add `math-token`, `math-mrow` etc for the box types that MathML needs. These values would have similar behavior as the existing [Layout-internal values](https://drafts.csswg.org/css-display/#layout-specific-display).  This is a simpler model that can be reasoned about.

**Comment by fantasai** [comment]
@MatsPalmgren Can you file the issue about adding math-internal values separately so we can keep this one focused on the top-level `math` value?

**Comment by fantasai** [comment]
So to summarize where I think we are here:

- RESOLVED: Add `math` as an inline-level `display` value.
- RESOLVED: Add `block math` for block-level math.
- Re-opened: Whether `display: math` on a non-MathML context element makes it behave like mrow or like `display: flow`.

@MatsPalmgren also posted issues about the mathml box fixup rules and about adding additional mathml display values, which should probably be pushed to their own independent issues.

**Comment by astearns** [comment]
@fred-wang this week’s meeting is at the bad time for Europeans. Shall we plan on taking this up on January 13? @MatsPalmgren would you be able to make next week’s call?

**Comment by bkardell** [comment]
> So to summarize where I think we are here:
> 
> * RESOLVED: Add `math` as an inline-level `display` value.
> * RESOLVED: Add `block math` for block-level math.
> * Re-opened: Whether `display: math` on a non-MathML context element makes it behave like mrow or like `display: flow`.
> 
> @MatsPalmgren also posted issues about the mathml box fixup rules and about adding additional internal mathml display values, which should probably be pushed to their own independent issues.

There are some other issues here - I'm not sure what is left on this issue specifically.. Is it literally just whether `display:math` on a non-MathML context should be more inline or block?

**Comment by fantasai** [comment]
@bkardell Yeah

**Comment by fantasai** [comment]
@bkardell Sorry, I misread. The question is whether `display: math` outside a MathML context should generate an mrow-like box.
